src/lib.rs

```diff
@@ -1227,44 +1266,47 @@ impl Device {
         }
     }
 
-    /// Creates a [`Buffer`].
-    pub fn create_buffer(&self, desc: &BufferDescriptor) -> Buffer {
-        let mut map_context = MapContext::new(desc.size);
-        if desc.mapped_at_creation {
-            map_context.initial_range = 0..desc.size;
-        }
-        Buffer {
-            context: Arc::clone(&self.context),
-            id: Context::device_create_buffer(&*self.context, &self.id, desc),
-            map_context: Mutex::new(map_context),
-            usage: desc.usage,
-        }
-    }
-
-    /// Creates a new buffer, maps it into host-visible memory, copies data from the given slice,
-    /// and finally unmaps it, returning a [`Buffer`].
-    pub fn create_buffer_with_data(&self, data: &[u8], usage: BufferUsage) -> Buffer {
-        let align = COPY_BUFFER_ALIGNMENT;
-        let unpadded_size = data.len() as u64;
-        let padding = (align - unpadded_size % align) % align;
-        let padded_size = unpadded_size + padding;
-
-        let buffer = self.create_buffer(&BufferDescriptor {
-            label: None,
-            size: padded_size,
-            usage,
-            mapped_at_creation: true,
-        });
-
-        let range =
-            Context::buffer_get_mapped_range_mut(&*self.context, &buffer.id, 0..padded_size);
-        range[0..unpadded_size as usize].copy_from_slice(&data);
-        for i in unpadded_size..padded_size {
-            range[i as usize] = 0;
+    /// Creates a [`Buffer`] with or without data to initialize it with.
+    pub fn create_buffer(&self, descriptor: &BufferInitDescriptor<'_, '_>) -> Buffer {
+        let wgt_descriptor = descriptor.to_buffer_descriptor();
```

Comment:
do we need `to_buffer_descriptor` at all? I think we end up matching on `descriptor.contents` the whole 3 times in the process, including this function!
Can we perhaps just do a single match in here without having a separate `to_buffer_descriptor` function?